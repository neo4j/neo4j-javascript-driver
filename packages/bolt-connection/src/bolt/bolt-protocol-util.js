/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { newError } from 'neo4j-driver-core'
import { ResultStreamObserver } from './stream-observers'

/**
 * @param {TxConfig} txConfig the auto-commit transaction configuration.
 * @param {function(error: string)} onProtocolError called when the txConfig is not empty.
 * @param {ResultStreamObserver} observer the response observer.
 */
function assertTxConfigIsEmpty (txConfig, onProtocolError = () => {}, observer) {
  if (txConfig && !txConfig.isEmpty()) {
    const error = newError(
      'Driver is connected to the database that does not support transaction configuration. ' +
        'Please upgrade to neo4j 3.5.0 or later in order to use this functionality'
    )

    // unsupported API was used, consider this a fatal error for the current connection
    onProtocolError(error.message)
    observer.onError(error)
    throw error
  }
}

/**
 * Asserts that the passed-in database name is empty.
 * @param {string} database
 * @param {fuction(err: String)} onProtocolError Called when it doesn't have database set
 */
function assertDatabaseIsEmpty (database, onProtocolError = () => {}, observer) {
  if (database) {
    const error = newError(
      'Driver is connected to the database that does not support multiple databases. ' +
        'Please upgrade to neo4j 4.0.0 or later in order to use this functionality'
    )

    // unsupported API was used, consider this a fatal error for the current connection
    onProtocolError(error.message)
    observer.onError(error)
    throw error
  }
}

export { assertDatabaseIsEmpty, assertTxConfigIsEmpty }
