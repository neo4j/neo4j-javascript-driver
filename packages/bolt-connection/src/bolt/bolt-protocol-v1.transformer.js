/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Node,
  newError,
  error,
  Relationship,
  UnboundRelationship,
  Path,
  toNumber,
  PathSegment
} from 'neo4j-driver-core'

import { structure } from '../packstream'
import { TypeTransformer } from './transformer'

const { PROTOCOL_ERROR } = error

const NODE = 0x4e
const NODE_STRUCT_SIZE = 3

const RELATIONSHIP = 0x52
const RELATIONSHIP_STRUCT_SIZE = 5

const UNBOUND_RELATIONSHIP = 0x72
const UNBOUND_RELATIONSHIP_STRUCT_SIZE = 3

const PATH = 0x50
const PATH_STRUCT_SIZE = 3

export function createNodeTransformer () {
  return new TypeTransformer({
    signature: NODE,
    isTypeInstance: object => object instanceof Node,
    toStructure: object => {
      throw newError(
        `It is not allowed to pass nodes in query parameters, given: ${object}`,
        PROTOCOL_ERROR
      )
    },
    fromStructure: struct => {
      structure.verifyStructSize('Node', NODE_STRUCT_SIZE, struct.size)

      const [identity, labels, properties] = struct.fields

      return new Node(identity, labels, properties)
    }
  })
}

export function createRelationshipTransformer () {
  return new TypeTransformer({
    signature: RELATIONSHIP,
    isTypeInstance: object => object instanceof Relationship,
    toStructure: object => {
      throw newError(
        `It is not allowed to pass relationships in query parameters, given: ${object}`,
        PROTOCOL_ERROR
      )
    },
    fromStructure: struct => {
      structure.verifyStructSize('Relationship', RELATIONSHIP_STRUCT_SIZE, struct.size)

      const [identity, startNodeIdentity, endNodeIdentity, type, properties] = struct.fields

      return new Relationship(identity, startNodeIdentity, endNodeIdentity, type, properties)
    }
  })
}

export function createUnboundRelationshipTransformer () {
  return new TypeTransformer({
    signature: UNBOUND_RELATIONSHIP,
    isTypeInstance: object => object instanceof UnboundRelationship,
    toStructure: object => {
      throw newError(
        `It is not allowed to pass unbound relationships in query parameters, given: ${object}`,
        PROTOCOL_ERROR
      )
    },
    fromStructure: struct => {
      structure.verifyStructSize(
        'UnboundRelationship',
        UNBOUND_RELATIONSHIP_STRUCT_SIZE,
        struct.size
      )

      const [identity, type, properties] = struct.fields

      return new UnboundRelationship(identity, type, properties)
    }
  })
}

export function createPathTransformer () {
  return new TypeTransformer({
    signature: PATH,
    isTypeInstance: object => object instanceof Path,
    toStructure: object => {
      throw newError(
        `It is not allowed to pass paths in query parameters, given: ${object}`,
        PROTOCOL_ERROR
      )
    },
    fromStructure: struct => {
      structure.verifyStructSize('Path', PATH_STRUCT_SIZE, struct.size)

      const [nodes, rels, sequence] = struct.fields

      const segments = []
      let prevNode = nodes[0]

      for (let i = 0; i < sequence.length; i += 2) {
        const nextNode = nodes[sequence[i + 1]]
        const relIndex = toNumber(sequence[i])
        let rel

        if (relIndex > 0) {
          rel = rels[relIndex - 1]
          if (rel instanceof UnboundRelationship) {
            // To avoid duplication, relationships in a path do not contain
            // information about their start and end nodes, that's instead
            // inferred from the path sequence. This is us inferring (and,
            // for performance reasons remembering) the start/end of a rel.
            rels[relIndex - 1] = rel = rel.bindTo(
              prevNode,
              nextNode
            )
          }
        } else {
          rel = rels[-relIndex - 1]
          if (rel instanceof UnboundRelationship) {
            // See above
            rels[-relIndex - 1] = rel = rel.bindTo(
              nextNode,
              prevNode
            )
          }
        }
        // Done hydrating one path segment.
        segments.push(new PathSegment(prevNode, rel, nextNode))
        prevNode = nextNode
      }
      return new Path(nodes[0], nodes[nodes.length - 1], segments)
    }
  })
}
