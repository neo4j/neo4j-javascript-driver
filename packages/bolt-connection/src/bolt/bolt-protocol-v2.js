/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import BoltProtocolV1 from './bolt-protocol-v1'
import v2 from '../packstream'
import { internal } from 'neo4j-driver-core'
import transformersFactories from './bolt-protocol-v2.transformer'
import Transformer from './transformer'

const {
  constants: { BOLT_PROTOCOL_V2 }
} = internal

export default class BoltProtocol extends BoltProtocolV1 {
  _createPacker (chunker) {
    return new v2.Packer(chunker)
  }

  _createUnpacker (disableLosslessIntegers, useBigInt) {
    return new v2.Unpacker(disableLosslessIntegers, useBigInt)
  }

  get transformer () {
    if (this._transformer === undefined) {
      this._transformer = new Transformer(Object.values(transformersFactories).map(create => create(this._config, this._log)), this._hydrationHooks, this._dehydrationHooks)
    }
    return this._transformer
  }

  get version () {
    return BOLT_PROTOCOL_V2
  }
}
