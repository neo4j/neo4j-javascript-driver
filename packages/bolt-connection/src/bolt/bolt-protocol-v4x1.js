/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import BoltProtocolV4 from './bolt-protocol-v4x0'
import RequestMessage from './request-message'
import { LoginObserver } from './stream-observers'
import { internal } from 'neo4j-driver-core'

import transformersFactories from './bolt-protocol-v4x1.transformer'
import Transformer from './transformer'

const {
  constants: { BOLT_PROTOCOL_V4_1 }
} = internal

export default class BoltProtocol extends BoltProtocolV4 {
  /**
   * @constructor
   * @param {Object} server the server informatio.
   * @param {Chunker} chunker the chunker.
   * @param {Object} packstreamConfig Packstream configuration
   * @param {boolean} packstreamConfig.disableLosslessIntegers if this connection should convert all received integers to native JS numbers.
   * @param {boolean} packstreamConfig.useBigInt if this connection should convert all received integers to native BigInt numbers.
   * @param {CreateResponseHandler} createResponseHandler Function which creates the response handler
   * @param {Logger} log the logger
   * @param {Object} serversideRouting
   *
   */
  constructor (
    server,
    chunker,
    packstreamConfig,
    createResponseHandler = () => null,
    log,
    onProtocolError,
    serversideRouting
  ) {
    super(
      server,
      chunker,
      packstreamConfig,
      createResponseHandler,
      log,
      onProtocolError
    )
    this._serversideRouting = serversideRouting
  }

  get version () {
    return BOLT_PROTOCOL_V4_1
  }

  get transformer () {
    if (this._transformer === undefined) {
      this._transformer = new Transformer(Object.values(transformersFactories).map(create => create(this._config)))
    }
    return this._transformer
  }

  initialize ({ userAgent, authToken, onError, onComplete } = {}) {
    const observer = new LoginObserver({
      onError: error => this._onLoginError(error, onError),
      onCompleted: metadata => this._onLoginCompleted(metadata, onComplete)
    })

    this.write(
      RequestMessage.hello(userAgent, authToken, this._serversideRouting),
      observer,
      true
    )

    return observer
  }
}
