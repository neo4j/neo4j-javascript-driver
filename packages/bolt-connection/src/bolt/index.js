/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import handshake from './handshake'
import create from './create'
import _BoltProtocol from './bolt-protocol-v4x3'
import _RawRoutingTable from './routing-table-raw'

export * from './stream-observers'

export const BoltProtocol = _BoltProtocol
export const RawRoutingTable = _RawRoutingTable

export default {
  handshake,
  create
}
