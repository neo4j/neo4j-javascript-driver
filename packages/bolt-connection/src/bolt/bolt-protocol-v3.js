/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import BoltProtocolV2 from './bolt-protocol-v2'
import RequestMessage from './request-message'
import { assertDatabaseIsEmpty, assertImpersonatedUserIsEmpty } from './bolt-protocol-util'
import {
  StreamObserver,
  LoginObserver,
  ResultStreamObserver,
  ProcedureRouteObserver
} from './stream-observers'
import { internal } from 'neo4j-driver-core'

const {
  bookmark: { Bookmark },
  constants: { BOLT_PROTOCOL_V3 },
  txConfig: { TxConfig }
} = internal

const CONTEXT = 'context'
const CALL_GET_ROUTING_TABLE = `CALL dbms.cluster.routing.getRoutingTable($${CONTEXT})`

const noOpObserver = new StreamObserver()

export default class BoltProtocol extends BoltProtocolV2 {
  get version () {
    return BOLT_PROTOCOL_V3
  }

  transformMetadata (metadata) {
    if ('t_first' in metadata) {
      // Bolt V3 uses shorter key 't_first' to represent 'result_available_after'
      // adjust the key to be the same as in Bolt V1 so that ResultSummary can retrieve the value
      metadata.result_available_after = metadata.t_first
      delete metadata.t_first
    }
    if ('t_last' in metadata) {
      // Bolt V3 uses shorter key 't_last' to represent 'result_consumed_after'
      // adjust the key to be the same as in Bolt V1 so that ResultSummary can retrieve the value
      metadata.result_consumed_after = metadata.t_last
      delete metadata.t_last
    }
    return metadata
  }

  initialize ({ userAgent, authToken, onError, onComplete } = {}) {
    const observer = new LoginObserver({
      onError: error => this._onLoginError(error, onError),
      onCompleted: metadata => this._onLoginCompleted(metadata, onComplete)
    })

    this.write(RequestMessage.hello(userAgent, authToken), observer, true)

    return observer
  }

  prepareToClose () {
    this.write(RequestMessage.goodbye(), noOpObserver, true)
  }

  beginTransaction ({
    bookmark,
    txConfig,
    database,
    impersonatedUser,
    mode,
    beforeError,
    afterError,
    beforeComplete,
    afterComplete
  } = {}) {
    const observer = new ResultStreamObserver({
      server: this._server,
      beforeError,
      afterError,
      beforeComplete,
      afterComplete
    })
    observer.prepareToHandleSingleResponse()

    // passing in a database name on this protocol version throws an error
    assertDatabaseIsEmpty(database, this._onProtocolError, observer)
    // passing impersonated user on this protocol version throws an error
    assertImpersonatedUserIsEmpty(impersonatedUser, this._onProtocolError, observer)

    this.write(
      RequestMessage.begin({ bookmark, txConfig, mode }),
      observer,
      true
    )

    return observer
  }

  commitTransaction ({
    beforeError,
    afterError,
    beforeComplete,
    afterComplete
  } = {}) {
    const observer = new ResultStreamObserver({
      server: this._server,
      beforeError,
      afterError,
      beforeComplete,
      afterComplete
    })
    observer.prepareToHandleSingleResponse()

    this.write(RequestMessage.commit(), observer, true)

    return observer
  }

  rollbackTransaction ({
    beforeError,
    afterError,
    beforeComplete,
    afterComplete
  } = {}) {
    const observer = new ResultStreamObserver({
      server: this._server,
      beforeError,
      afterError,
      beforeComplete,
      afterComplete
    })
    observer.prepareToHandleSingleResponse()

    this.write(RequestMessage.rollback(), observer, true)

    return observer
  }

  run (
    query,
    parameters,
    {
      bookmark,
      txConfig,
      database,
      impersonatedUser,
      mode,
      beforeKeys,
      afterKeys,
      beforeError,
      afterError,
      beforeComplete,
      afterComplete,
      flush = true
    } = {}
  ) {
    const observer = new ResultStreamObserver({
      server: this._server,
      beforeKeys,
      afterKeys,
      beforeError,
      afterError,
      beforeComplete,
      afterComplete
    })

    // passing in a database name on this protocol version throws an error
    assertDatabaseIsEmpty(database, this._onProtocolError, observer)
    // passing impersonated user on this protocol version throws an error
    assertImpersonatedUserIsEmpty(impersonatedUser, this._onProtocolError, observer)

    this.write(
      RequestMessage.runWithMetadata(query, parameters, {
        bookmark,
        txConfig,
        mode
      }),
      observer,
      false
    )
    this.write(RequestMessage.pullAll(), observer, flush)

    return observer
  }

  /**
   * Request routing information
   *
   * @param {Object} param -
   * @param {object} param.routingContext The routing context used to define the routing table.
   *  Multi-datacenter deployments is one of its use cases
   * @param {string} param.databaseName The database name
   * @param {Bookmark} params.sessionContext.bookmark The bookmark used for request the routing table
   * @param {string} params.sessionContext.mode The session mode
   * @param {string} params.sessionContext.database The database name used on the session
   * @param {function()} params.sessionContext.afterComplete The session param used after the session closed
   * @param {function(err: Error)} param.onError
   * @param {function(RawRoutingTable)} param.onCompleted
   * @returns {RouteObserver} the route observer
   */
  requestRoutingInformation ({
    routingContext = {},
    sessionContext = {},
    onError,
    onCompleted
  }) {
    const resultObserver = this.run(
      CALL_GET_ROUTING_TABLE,
      { [CONTEXT]: routingContext },
      { ...sessionContext, txConfig: TxConfig.empty() }
    )

    return new ProcedureRouteObserver({
      resultObserver,
      onProtocolError: this._onProtocolError,
      onError,
      onCompleted
    })
  }
}
