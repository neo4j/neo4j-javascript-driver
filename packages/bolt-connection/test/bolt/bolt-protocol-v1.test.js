/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import BoltProtocolV1 from '../../src/bolt/bolt-protocol-v1'
import RequestMessage from '../../src/bolt/request-message'
import { internal } from 'neo4j-driver-core'
import utils from '../test-utils'
import { LoginObserver } from '../../src/bolt/stream-observers'

const WRITE = 'WRITE'

const {
  bookmark: { Bookmark },
  txConfig: { TxConfig }
} = internal

describe('#unit BoltProtocolV1', () => {
  beforeEach(() => {
    expect.extend(utils.matchers)
  })

  it('should not change metadata', () => {
    const metadata = {
      result_available_after: 1,
      result_consumed_after: 2,
      t_first: 3,
      t_last: 4
    }
    const protocol = new BoltProtocolV1(
      new utils.MessageRecordingConnection(),
      null,
      false
    )

    const transformedMetadata = protocol.transformMetadata(metadata)

    expect(transformedMetadata).toEqual({
      result_available_after: 1,
      result_consumed_after: 2,
      t_first: 3,
      t_last: 4
    })
  })

  it('should initialize the connection', () => {
    const recorder = new utils.MessageRecordingConnection()
    const protocol = utils.spyProtocolWrite(
      new BoltProtocolV1(recorder, null, false)
    )

    const onError = _error => {}
    const onComplete = () => {}
    const clientName = 'js-driver/1.2.3'
    const authToken = { username: 'neo4j', password: 'secret' }

    const observer = protocol.initialize({
      userAgent: clientName,
      authToken,
      onError,
      onComplete
    })

    expect(observer).toBeTruthy()
    expect(observer instanceof LoginObserver).toBeTruthy()

    protocol.verifyMessageCount(1)
    expect(protocol.messages[0]).toBeMessage(
      RequestMessage.init(clientName, authToken)
    )
    expect(protocol.observers).toEqual([observer])
    expect(protocol.flushes).toEqual([true])
  })

  it('should run a query', () => {
    const recorder = new utils.MessageRecordingConnection()
    const protocol = utils.spyProtocolWrite(
      new BoltProtocolV1(recorder, null, false)
    )

    const query = 'RETURN $x, $y'
    const parameters = { x: 'x', y: 'y' }
    const observer = protocol.run(query, parameters, {
      bookmark: Bookmark.empty(),
      txConfig: TxConfig.empty(),
      mode: WRITE
    })

    protocol.verifyMessageCount(2)

    expect(protocol.messages[0]).toBeMessage(
      RequestMessage.run(query, parameters)
    )
    expect(protocol.messages[1]).toBeMessage(RequestMessage.pullAll())
    expect(protocol.observers).toEqual([observer, observer])
    expect(protocol.flushes).toEqual([false, true])
  })

  it('should reset the connection', () => {
    const recorder = new utils.MessageRecordingConnection()
    const protocol = utils.spyProtocolWrite(
      new BoltProtocolV1(recorder, null, false)
    )

    const observer = protocol.reset()

    protocol.verifyMessageCount(1)
    expect(protocol.messages[0]).toBeMessage(RequestMessage.reset())
    expect(protocol.observers).toEqual([observer])
    expect(protocol.flushes).toEqual([true])
  })

  it('should begin a transaction', () => {
    const recorder = new utils.MessageRecordingConnection()
    const protocol = utils.spyProtocolWrite(
      new BoltProtocolV1(recorder, null, false)
    )

    const bookmark = new Bookmark('neo4j:bookmark:v1:tx42')

    const observer = protocol.beginTransaction({
      bookmark: bookmark,
      txConfig: TxConfig.empty(),
      mode: WRITE
    })

    protocol.verifyMessageCount(2)

    expect(protocol.messages[0]).toBeMessage(
      RequestMessage.run('BEGIN', bookmark.asBeginTransactionParameters())
    )
    expect(protocol.messages[1]).toBeMessage(RequestMessage.pullAll())
    expect(protocol.observers).toEqual([observer, observer])
    expect(protocol.flushes).toEqual([false, false])
  })

  it('should commit a transaction', () => {
    const recorder = new utils.MessageRecordingConnection()
    const protocol = utils.spyProtocolWrite(
      new BoltProtocolV1(recorder, null, false)
    )

    const observer = protocol.commitTransaction()

    protocol.verifyMessageCount(2)

    expect(protocol.messages[0]).toBeMessage(RequestMessage.run('COMMIT', {}))
    expect(protocol.messages[1]).toBeMessage(RequestMessage.pullAll())
    expect(protocol.observers).toEqual([observer, observer])
    expect(protocol.flushes).toEqual([false, true])
  })

  it('should rollback a transaction', () => {
    const recorder = new utils.MessageRecordingConnection()
    const protocol = utils.spyProtocolWrite(
      new BoltProtocolV1(recorder, null, false)
    )

    const observer = protocol.rollbackTransaction()

    protocol.verifyMessageCount(2)

    expect(protocol.messages[0]).toBeMessage(RequestMessage.run('ROLLBACK', {}))
    expect(protocol.messages[1]).toBeMessage(RequestMessage.pullAll())
    expect(protocol.observers).toEqual([observer, observer])
    expect(protocol.flushes).toEqual([false, true])
  })

  it('should return correct bolt version number', () => {
    const protocol = new BoltProtocolV1(null, null, false)

    expect(protocol.version).toBe(1)
  })

  describe('Bolt V3', () => {
    /**
     * @param {function(protocol: BoltProtocolV1)} fn
     */
    function verifyError (fn) {
      const recorder = new utils.MessageRecordingConnection()
      const protocol = new BoltProtocolV1(recorder, null, false)

      expect(() => fn(protocol)).toThrowError(
        'Driver is connected to the database that does not support transaction configuration. ' +
          'Please upgrade to neo4j 3.5.0 or later in order to use this functionality'
      )
    }

    describe('beginTransaction', () => {
      function verifyBeginTransaction (txConfig) {
        verifyError(protocol => protocol.beginTransaction({ txConfig }))
      }

      it('should throw error when txConfig.timeout is set', () => {
        verifyBeginTransaction(new TxConfig({ timeout: 5000 }))
      })

      it('should throw error when txConfig.metadata is set', () => {
        verifyBeginTransaction(new TxConfig({ metadata: { x: 1, y: true } }))
      })

      it('should throw error when txConfig is set', () => {
        verifyBeginTransaction(
          new TxConfig({ timeout: 5000, metadata: { x: 1, y: true } })
        )
      })
    })

    describe('run', () => {
      function verifyRun (txConfig) {
        verifyError((protocol, _observer) =>
          protocol.run('query', {}, { txConfig })
        )
      }

      it('should throw error when txConfig.timeout is set', () => {
        verifyRun(new TxConfig({ timeout: 5000 }))
      })

      it('should throw error when txConfig.metadata is set', () => {
        verifyRun(new TxConfig({ metadata: { x: 1, y: true } }))
      })

      it('should throw error when txConfig is set', () => {
        verifyRun(new TxConfig({ timeout: 5000, metadata: { x: 1, y: true } }))
      })
    })
  })

  describe('Bolt V4', () => {
    /**
     * @param {function(protocol: BoltProtocolV1)} fn
     */
    function verifyError (fn) {
      const recorder = new utils.MessageRecordingConnection()
      const protocol = new BoltProtocolV1(recorder, null, false)

      expect(() => fn(protocol)).toThrowError(
        'Driver is connected to the database that does not support multiple databases. ' +
          'Please upgrade to neo4j 4.0.0 or later in order to use this functionality'
      )
    }

    describe('beginTransaction', () => {
      function verifyBeginTransaction (database) {
        verifyError(protocol => protocol.beginTransaction({ database }))
      }

      it('should throw error when database is set', () => {
        verifyBeginTransaction('test')
      })
    })

    describe('run', () => {
      function verifyRun (database) {
        verifyError(protocol => protocol.run('query', {}, { database }))
      }

      it('should throw error when database is set', () => {
        verifyRun('test')
      })
    })
  })

  describe('Bolt v4.4', () => {
    /**
     * @param {string} impersonatedUser The impersonated user.
     * @param {function(protocol: BoltProtocolV1)} fn 
     */
    function verifyImpersonationNotSupportedErrror (impersonatedUser, fn) {
      const recorder = new utils.MessageRecordingConnection()
      const protocol = new BoltProtocolV1(recorder, null, false)

      expect(() => fn(protocol)).toThrowError(
        'Driver is connected to the database that does not support user impersonation. ' +
          'Please upgrade to neo4j 4.4.0 or later in order to use this functionality. ' +
          `Trying to impersonate ${impersonatedUser}.`
      )
    }

    describe('beginTransaction', () => {
      function verifyBeginTransaction(impersonatedUser) {
        verifyImpersonationNotSupportedErrror(
          impersonatedUser,
          protocol => protocol.beginTransaction({ impersonatedUser }))
      }

      it('should throw error when impersonatedUser is set', () => {
        verifyBeginTransaction('test')
      })
    })

    describe('run', () => {
      function verifyRun (impersonatedUser) {
        verifyImpersonationNotSupportedErrror(
          impersonatedUser,
          protocol => protocol.run('query', {}, { impersonatedUser }))
      }

      it('should throw error when impersonatedUser is set', () => {
        verifyRun('test')
      })
    })
  })

  describe('unpacker configuration', () => {
    test.each([
      [false, false],
      [false, true],
      [true, false],
      [true, true]
    ])(
      'should create unpacker with disableLosslessIntegers=%p and useBigInt=%p',
      (disableLosslessIntegers, useBigInt) => {
        const protocol = new BoltProtocolV1(null, null, {
          disableLosslessIntegers,
          useBigInt
        })
        expect(protocol._unpacker._disableLosslessIntegers).toBe(
          disableLosslessIntegers
        )
        expect(protocol._unpacker._useBigInt).toBe(useBigInt)
      }
    )
  })
})
