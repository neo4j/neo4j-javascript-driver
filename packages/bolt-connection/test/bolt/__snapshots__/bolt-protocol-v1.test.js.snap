// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#unit BoltProtocolV1 .packable() should pack not pack graph types (Node) 1`] = `"It is not allowed to pass nodes in query parameters, given: (c:a {a:\\"b\\"})"`;

exports[`#unit BoltProtocolV1 .packable() should pack not pack graph types (Path) 1`] = `"It is not allowed to pass paths in query parameters, given: [object Object]"`;

exports[`#unit BoltProtocolV1 .packable() should pack not pack graph types (PathSegment) 1`] = `"Cannot read properties of null (reading 'writeUInt8')"`;

exports[`#unit BoltProtocolV1 .packable() should pack not pack graph types (Relationship) 1`] = `"It is not allowed to pass relationships in query parameters, given: (e)-[:a {b:\\"c\\"}]->(f)"`;

exports[`#unit BoltProtocolV1 .packable() should pack not pack graph types (UnboundRelationship) 1`] = `"It is not allowed to pass unbound relationships in query parameters, given: -[:a {b:\\"c\\"}]->"`;

exports[`#unit BoltProtocolV1 .packable() should pack types introduced afterwards as Map (Date) 1`] = `
Object {
  "day": 1,
  "month": 1,
  "year": 1,
}
`;

exports[`#unit BoltProtocolV1 .packable() should pack types introduced afterwards as Map (DateTime) 1`] = `
Object {
  "day": 1,
  "hour": 1,
  "minute": 1,
  "month": 1,
  "nanosecond": 1,
  "second": 1,
  "timeZoneOffsetSeconds": 1,
  "year": 1,
}
`;

exports[`#unit BoltProtocolV1 .packable() should pack types introduced afterwards as Map (Duration) 1`] = `
Object {
  "days": 1,
  "months": 1,
  "nanoseconds": Integer {
    "high": 0,
    "low": 1,
  },
  "seconds": Integer {
    "high": 0,
    "low": 1,
  },
}
`;

exports[`#unit BoltProtocolV1 .packable() should pack types introduced afterwards as Map (LocalDateTime) 1`] = `
Object {
  "day": 1,
  "hour": 1,
  "minute": 1,
  "month": 1,
  "nanosecond": 1,
  "second": 1,
  "year": 1,
}
`;

exports[`#unit BoltProtocolV1 .packable() should pack types introduced afterwards as Map (LocalTime) 1`] = `
Object {
  "hour": 1,
  "minute": 1,
  "nanosecond": 1,
  "second": 1,
}
`;

exports[`#unit BoltProtocolV1 .packable() should pack types introduced afterwards as Map (Point2D) 1`] = `
Object {
  "srid": 1,
  "x": 1,
  "y": 1,
}
`;

exports[`#unit BoltProtocolV1 .packable() should pack types introduced afterwards as Map (Point3D) 1`] = `
Object {
  "srid": 1,
  "x": 1,
  "y": 1,
  "z": 1,
}
`;

exports[`#unit BoltProtocolV1 .packable() should pack types introduced afterwards as Map (Time) 1`] = `
Object {
  "hour": 1,
  "minute": 1,
  "nanosecond": 1,
  "second": 1,
  "timeZoneOffsetSeconds": 1,
}
`;

exports[`#unit BoltProtocolV1 .unpack() should not unpack graph types with wrong size(Node with less fields) 1`] = `"Wrong struct size for Node, expected 3 but was 2"`;

exports[`#unit BoltProtocolV1 .unpack() should not unpack graph types with wrong size(Node with more fields) 1`] = `"Wrong struct size for Node, expected 3 but was 4"`;

exports[`#unit BoltProtocolV1 .unpack() should not unpack graph types with wrong size(Path with less fields) 1`] = `"Wrong struct size for Path, expected 3 but was 2"`;

exports[`#unit BoltProtocolV1 .unpack() should not unpack graph types with wrong size(Path with more fields) 1`] = `"Wrong struct size for Path, expected 3 but was 4"`;

exports[`#unit BoltProtocolV1 .unpack() should not unpack graph types with wrong size(Relationship with less fields) 1`] = `"Wrong struct size for Relationship, expected 5 but was 4"`;

exports[`#unit BoltProtocolV1 .unpack() should not unpack graph types with wrong size(Relationship with more fields) 1`] = `"Wrong struct size for Relationship, expected 5 but was 8"`;

exports[`#unit BoltProtocolV1 .unpack() should not unpack graph types with wrong size(UnboundRelationship with less fields) 1`] = `"Wrong struct size for UnboundRelationship, expected 3 but was 2"`;

exports[`#unit BoltProtocolV1 .unpack() should not unpack graph types with wrong size(UnboundRelationship with more fields) 1`] = `"Wrong struct size for UnboundRelationship, expected 3 but was 4"`;
